API/ev3_timer.c

```diff
@@ -12,268 +16,552 @@
  * You should have received a copy of the GNU General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/
  *
- * The Initial Developer of this code is John Hansen.
- * Portions created by John Hansen are Copyright (C) 2009-2013 John Hansen.
+ * The Initial Developer of this code is Jakub Vanek.
+ * Portions created by Jakub Vanek are Copyright (C) 2020 Jakub Vanek.
  * All Rights Reserved.
  *
+ * ----------------------------------------------------------------------------
+ *
+ * \author Jakub Vanek (linuxtardis_at_gmail.com)
+ * \date 2020-02-11
+ * \version 1
  */
 
-#include "ev3_timer.h"
+#include "ev3.h"
+#include "ev3_timer.private.h"
 
-#ifndef DISABLE_TIMERS
+////////////////////////////////////////////////////////////////
+// STATIC DATA
 
-static unsigned long long msTimers[NUM_MS_TIMERS];
-static unsigned long long usTimers[NUM_US_TIMERS];
-static unsigned long long csTimers[NUM_CS_TIMERS];
+//! Global data of the timer module.
+struct TimerControlBlock tcb = {
+        .initialized = false
+};
 
-#define TIMER_INTERVAL 10000 // 10 ms
-#define TIMER_10MS     1
-#define TIMER_50MS     5
-#define TIMER_100MS    10
-#define TIMER_250MS    25
-#define TIMER_500MS    50
-#define TIMER_1SEC     100
-#define MAX_CALLBACKS  4
 
-static TimerCallback callBack10ms[MAX_CALLBACKS];
-static TimerCallback callBack50ms[MAX_CALLBACKS];
-static TimerCallback callBack100ms[MAX_CALLBACKS];
-static TimerCallback callBack250ms[MAX_CALLBACKS];
-static TimerCallback callBack500ms[MAX_CALLBACKS];
-static TimerCallback callBack1s[MAX_CALLBACKS];
+////////////////////////////////////////////////////////////////
+// IMPLEMENTATION
 
-static int callBack10ms_count;
-static int callBack50ms_count;
-static int callBack100ms_count;
-static int callBack250ms_count;
-static int callBack500ms_count;
-static int callBack1s_count;
+// INITIALIZATION
 
-static unsigned long timer_interval_mutliplier = 1;
+bool TimerInitialized()
+{
+    __sync_synchronize();
+    bool result = tcb.initialized;

```

Comment:
Using `__sync_lock_test_and_set` and `__sync_lock_release` should indeed be better here. We cannot use `__atomic_*` functions because that would (probably?) break GCC 4.3 which is still supported.

Reply:
Ye, I didn't mean C11 GCC primitive `__atomic`. `__sync_lock_test_and_set` is fine.

Reply:
I had to use the CAS operation to perform this, as I'm haven't realized the release/acquire semantics of __sync lock/release. Resolved in 39603c1ae5f0993508ae3172b6581354257615ef

Reply:
I realize now what I think was your intention: initialization happens in the construtor, so thread-safety is a non issue as we haven't entered `main` yet. I agree with that argument, but as you are nearly there, thread safety doesn't hurt.